src/main/scala/zio/json/encoder.scala

```diff
@@ -4,42 +4,108 @@ import scala.annotation._
+   * encoder and the specified encoder.
+   */
+  final def both[B](that: => JsonEncoder[B]): JsonEncoder[(A, B)] = JsonEncoder.tuple2(self, that)
 
-  // scalaz-deriving style Contravariant combinators
-  final def narrow[B <: A]: Encoder[B] = self.asInstanceOf[Encoder[B]]
-  final def contramap[B](f: B => A): Encoder[B] = new Encoder[B] {
+  /**
+   * Returns a new encoder that is capable of encoding a user-defined value, which is create from
+   * a tuple of the values of this encoder and the specified encoder, from the specified user-
+   * defined function.
+   */
+  final def bothWith[B, C](that: => JsonEncoder[B])(f: C => (A, B)): JsonEncoder[C] = self.both(that).contramap(f)
+
+  /**
+   * Returns a new encoder, with a new input type, which can be transformed to the old input type
+   * by the specified user-defined function.
+   */
+  final def contramap[B](f: B => A): JsonEncoder[B] = new JsonEncoder[B] {
     override def unsafeEncode(b: B, indent: Option[Int], out: java.io.Writer): Unit =
       self.unsafeEncode(f(b), indent, out)
     override def isNothing(b: B): Boolean = self.isNothing(f(b))
   }
-  final def xmap[B](f: A => B, g: B => A): Encoder[B] = contramap(g)
 
-  def unsafeEncode(a: A, indent: Option[Int], out: java.io.Writer): Unit
+  /**
+   * Returns a new encoder that can accepts an `Either[A, B]` to either, and uses either this
+   * encoder or the specified encoder to encode the two different types of values.
+   */
+  final def either[B](that: => JsonEncoder[B]): JsonEncoder[Either[A, B]] = JsonEncoder.either[A, B](self, that)
+
+  /**
+   * Returns a new encoder with a new input type, which can be transformed to either the input
+   * type of this encoder, or the input type of the specified encoder, using the user-defined
+   * transformation function.
+   */
+  final def eitherWith[B, C](that: => JsonEncoder[B])(f: C => Either[A, B]): JsonEncoder[C] =
+    self.either(that).contramap(f)
+
+  /**
+   * Encodes the specified value into a JSON string, with the specified indentation level.
+   */
+  final def encodeJson(a: A, indent: Option[Int]): String = {
+    val writer = new zio.json.internal.FastStringWriter(64)
+    unsafeEncode(a, indent, writer)
+    writer.toString
+  }
+
+  /**
+   * Encodes the specified value into a character stream.
+   *
+   * TODO: Use Take so we can push errors into the stream.
+   */
+  final def encodeJsonStream(a: A, indent: Option[Int]): ZStream[Blocking, Nothing, Char] =
+    ZStream.unwrapManaged {
+      (for {
+        runtime <- ZIO.runtime[Any].toManaged_
+        queue   <- Queue.bounded[Chunk[Char]](1).toManaged_
+        writer <- ZManaged.fromAutoCloseable {
+                   ZIO.effectTotal {
+                     new java.io.BufferedWriter(new Writer {
+                       override def write(buffer: Array[Char], offset: Int, len: Int): Unit =
+                         runtime.unsafeRun(queue.offer(Chunk.fromArray(buffer).drop(offset).take(len)))
```

Comment:
The underlying array is modified after calling `Chunk.fromArray`, leading to data corruption

Reply:
Good catch!